### PR TITLE
Validation loss

### LINKED ISSUE #60: [FEATURE] Add validation loss calculation
**Is your feature request related to a problem? Please describe.**
Nope

**Describe the solution you'd like**
Add new metric which is validation loss, which in effect will also possible to add reducelronplateau scheduler. And i think in most paper they always show relationship between training and validation loss

**Describe alternatives you've considered**
Possible a new dedicated class for this or make a static method under base validator, subject to discuss

### PR DESCRIPTION
**Pull request by alphinside**
## Type of changes

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?
No mechanism to calculate loss on validation data

- close #60 

## What is the new behavior?

- Add validation loss to experiment logger
-
-

## Checklist

- [x] Unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have filled the `[Unreleased]` [Changelog](https://github.com/nodefluxio/vortex/blob/master/CHANGELOG.md)

### CODE REVIEW

File: experiments/configs/shufflenetv2x100_classification_cifar10.yml [hunk 1]
```diff
@@ -38,7 +38,7 @@ model: {
     freeze_backbone: False
   },
   preprocess_args: {
-    input_size: 32,
+    input_size: 224,
```
Comment (triwahyuu) [inline experiments/configs/shufflenetv2x100_classification_cifar10.yml:41]: you do know that cifar's original image size is 32px right?

**Review by triwahyuu — COMMENTED** [review]
is there any other way so that we could validate only once, rather than twice for each metrics and loss? probably get the output of model from predictor using [forward hook](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_forward_hook)?

**Comment by alphinside** [comment]
> is there any other way so that we could validate only once, rather than twice for each metrics and loss? probably get the output of model from predictor using [forward hook](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_forward_hook)?

currently i havent found any good design implementation about that, maybe imrpoved in the future issue i think

File: experiments/configs/shufflenetv2x100_classification_cifar10.yml (hunk 1 above)
Comment (alphinside) [reply experiments/configs/shufflenetv2x100_classification_cifar10.yml:41]: sorry i didn't aware that i modified this, will check it

**Comment by triwahyuu** [comment]
so, the current problem to not use the original basevalidator class is because the use of predictor right? why not use a forward hook?

**Comment by alphinside** [comment]
> so, the current problem to not use the original basevalidator class is because the use of predictor right? why not use a forward hook?

havent heard of this method before, so currently i have no idea yet on how to implement this and it's impact on our predictor design

**Comment by triwahyuu** [comment]
you can learn it in this post https://blog.paperspace.com/pytorch-hooks-gradient-clipping-debugging

**Comment by triwahyuu** [comment]
forward hooks basically is a function that is executed after the forward, so you save the output value inside the hook function to be used later on

File: src/development/vortex/development/core/engine/validator/base_validator.py [hunk 2]
```diff
@@ -382,44 +395,26 @@ def __call__(self, *args, **kwargs):
...
+                    try:
+                        raw_results = self.raw_model_output
+                        batch = list(zip(image,targets))
+                        _,targets = self.collate_fn(batch)
+                        targets = targets.to(raw_results.device)
+                        batch_loss = self.criterion(raw_results,targets)
+                        val_loss += batch_loss.detach()
+                        self.raw_model_output = None
+                    except Exception as e:
+                        print(e)
```
Comment (triwahyuu) [inline src/development/vortex/development/core/engine/validator/base_validator.py:409]: probably better to not print anything here, because if it fails once, it will always fail for all the dataset. have a warning in the end of validation would be better

Comment (alphinside) [reply src/development/vortex/development/core/engine/validator/base_validator.py:409]: Noted
